File: Java/Linked_List_05.java
// Detecting Loop in a Linked List

class Linked_List {
    Node head;
    int size;

    class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    public void Add_Last(int data) {
        size++;
        Node newNode = new Node(data);

        if (head == null) {
            head = newNode;
            return;
        }
        
        Node curr = head;
        while (curr.next != null) {
            curr = curr.next;
        }
        curr.next = newNode;
    }

    public void Print_Linked_List() {
        if (head == null) {
            System.out.println("List is empty");
            return;
        }

        Node curr = head;
        while (curr != null) {
            System.out.print(curr.data + " -> ");
            curr = curr.next;
        }
        System.out.println("NULL");
    }

    public boolean Detect_cycle(Node head){
        Node slow=head;
        Node fast=head;

        while(fast!=null && fast.next!=null){
            slow=slow.next;
            fast=fast.next.next;

            if(slow==fast){
                return true;
            }
        }

        return false;
    }
}

public class Linked_List_05 {
    public static void main(String[] args){
        Linked_List list = new Linked_List();

        list.Add_Last(0);
        list.Add_Last(1);
        list.Add_Last(2);
        list.Add_Last(3);
        list.Add_Last(4);
        list.Add_Last(5);

        list.Print_Linked_List();

        System.out.println(list.Detect_cycle(list.head));
    }
}

File: C++/Linked_list_05.cpp
// In this code we will see how to delete a node from different positions of a singly linked list

#include<iostream>

using namespace std;

class Node{
    public:
    int val;
    Node* next;

    Node(int data){
        val=data;
        next=NULL;
    }
};

void insert_at_head(Node* &head,int val){
    Node* new_node=new Node(val);

    new_node->next=head;
    head=new_node;
}

void delete_at_head(Node* &head){
    Node* temp=head;
    head=head->next;
    free(temp);
}

void delete_at_tail(Node* &head){
    Node* temp=head;

    while(temp->next->next!=NULL){
        temp=temp->next;
    }

    Node* tail_node=temp->next;
    temp->next=NULL;
    free(tail_node);
    
}

void delete_at_position(Node* &head,int position){
    Node* temp=head;

    int pos=1;
    while(pos!=position-1){
        temp=temp->next;
        pos++;
    }

    Node* target_node=temp->next;
    temp->next=temp->next->next;
    free(target_node);
}

void print_linked_list(Node* head){
    Node* temp=head;

    while(temp!=NULL){
        cout << temp->val << " -> ";
        temp=temp->next;
    }
    cout << "Null" << endl;
}

int main(){
    Node* linked_list=new Node(5);

    insert_at_head(linked_list,4);
    insert_at_head(linked_list,3);
    insert_at_head(linked_list,2);
    insert_at_head(linked_list,1);
    insert_at_head(linked_list,0);

    print_linked_list(linked_list);
    
    delete_at_head(linked_list);
    
    print_linked_list(linked_list);

    delete_at_tail(linked_list);
    
    print_linked_list(linked_list);

    delete_at_position(linked_list,2);
    
    print_linked_list(linked_list);

    return 0;
}